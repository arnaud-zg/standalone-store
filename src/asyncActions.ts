import { AnyAction, Middleware, StoreCreator } from 'redux'
import { ActionCreator, isActionOf } from 'typesafe-actions'
import { StandaloneStore } from '.'

interface IResolve<TState> {
  action: AnyAction
  state: TState
}

type ConfigureStore = ({
  middlewares,
}: {
  middlewares: Middleware[]
}) => ReturnType<StoreCreator>

export const dispatchActionsAndWaitResponse = <TState, TSelectorResponse>({
  actionsDispatch,
  actionCreatorsResolve,
  configureStore,
  selector,
}: {
  configureStore: ConfigureStore
  actionsDispatch: AnyAction[]
  actionCreatorsResolve: ActionCreator[]
  selector: (state: TState) => TSelectorResponse
}) => {
  const standaloneStore = new StandaloneStore<TState>({ configureStore })

  return !actionsDispatch.length
    ? Promise.reject('You should at least give one action.')
    : new Promise((resolve: ({ action, state }: IResolve<TState>) => void) => {
        actionCreatorsResolve.forEach(actionResolve => {
          standaloneStore.subscribe((action, state) => {
            if (isActionOf(actionResolve, action)) {
              resolve({ action, state })
            }
          })
        })
        actionsDispatch.forEach(action => {
          standaloneStore.dispatchAction(action)
        })
      }).then(({ state }: { state: TState }) => selector(state))
}
