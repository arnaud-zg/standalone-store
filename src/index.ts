export * from './asyncActions'
export * from './standaloneStore'
