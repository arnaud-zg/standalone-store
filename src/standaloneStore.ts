import { AnyAction, StoreCreator } from 'redux'

type Listener<TStoreState> = (
  action: AnyAction | null,
  state: TStoreState
) => void

export class StandaloneStore<TStoreState> {
  private listeners: Array<Listener<TStoreState>>
  private store: ReturnType<StoreCreator>
  private lastAction: AnyAction | null

  constructor({ store }: { store: ReturnType<StoreCreator> }) {
    this.lastAction = null
    this.listeners = []
    this.store = store

    this.store.subscribe(this.storeUpdate)
  }

  dispatchAction = (action: AnyAction) => {
    this.lastAction = action
    this.store.dispatch(action)
  }

  storeUpdate = () => {
    const state = this.store.getState() as TStoreState

    if (this.listeners.length) {
      this.listeners.forEach(listener => {
        listener(this.lastAction, state)
      })
    }
  }

  subscribe(listener: Listener<TStoreState>) {
    this.listeners.push(listener)
  }
}
