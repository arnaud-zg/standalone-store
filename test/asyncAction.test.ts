import { configureStore } from 'ts-foursquare'
import { dispatchActionsAndWaitResponse } from '../src'

describe('asyncAction', () => {
  it(`should throw an error`, done => {
    dispatchActionsAndWaitResponse({
      actionsDispatch: [],
      actionCreatorsResolve: [],
      configureStore,
      selector: () => {},
    }).catch(error => {
      expect(error).toMatchSnapshot()
      done()
    })
  })
})
