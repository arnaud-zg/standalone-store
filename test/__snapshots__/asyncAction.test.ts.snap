// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncAction should throw an error 1`] = `"You should at least give one action."`;
